File: Kalorie/Kalorie/Core/Networking/FireStone/FavouriteFoodDTO.swift
//
//  FavouriteFoodDTO.swift
//  Kalorie
//
//  Created by Josef Antoni on 12.08.2026.
//

import Foundation

struct FavouriteFoodDTO: Codable {

    // MARK: - Properties

    let id: String
    let czName: String
    let engName: String
    let weight: Double
    let date: TimeInterval
    let energyKJ: Double?
    let caloriesPerHundredGrams: Double
    let fat: Double
    let fatSaturated: Double?
    let fatUnsaturatedFattyAcids: Double
    let carbohydrate: Double
    let carbohydratePureSugar: Double
    let fiber: Double?
    let protein: Double
    let salt: Double
    let favouritedAt: TimeInterval

    // MARK: - Coding keys

    enum CodingKeys: String, CodingKey {
        case id, weight, date, fat, carbohydrate, protein, salt, fiber
        case czName = "cz_name"
        case engName = "eng_name"
        case energyKJ = "energy_kj"
        case caloriesPerHundredGrams = "calories_per_hundred_grams"
        case fatSaturated = "fat_saturated"
        case fatUnsaturatedFattyAcids = "fat_unsaturated_fatty_acids"
        case carbohydratePureSugar = "carbohydrate_pure_sugar"
        case favouritedAt = "favourited_at"
    }

    // MARK: - Init

    init(item: FoodItemDomain, favouritedAt: Date) {
        id = item.id
        czName = item.czName
        engName = item.engName
        weight = item.weight
        date = item.date.timeIntervalSince1970
        energyKJ = item.energyKJ
        caloriesPerHundredGrams = item.caloriesPerHundredGrams
        fat = item.fat
        fatSaturated = item.fatSaturated
        fatUnsaturatedFattyAcids = item.fatUnsaturatedFattyAcids
        carbohydrate = item.carbohydrate
        carbohydratePureSugar = item.carbohydratePureSugar
        fiber = item.fiber
        protein = item.protein
        salt = item.salt
        self.favouritedAt = favouritedAt.timeIntervalSince1970
    }

    // MARK: - Functions

    func asDomain() -> FoodItemDomain {
        FoodItemDomain(
            id: id,
            czName: czName,
            engName: engName,
            weight: weight,
            date: Date(timeIntervalSince1970: date),
            energyKJ: energyKJ ?? 0,
            caloriesPerHundredGrams: caloriesPerHundredGrams,
            fat: fat,
            fatSaturated: fatSaturated ?? 0,
            fatUnsaturatedFattyAcids: fatUnsaturatedFattyAcids,
            carbohydrate: carbohydrate,
            carbohydratePureSugar: carbohydratePureSugar,
            fiber: fiber ?? 0,
            protein: protein,
            salt: salt
        )
    }
}

File: Android/app/src/main/java/antoni/kalorie/core/networking/FavouriteFoodDTO.kt
package antoni.kalorie.core.networking

import antoni.kalorie.core.models.FoodItemDomain
import antoni.kalorie.core.models.FoodItemKind
import antoni.kalorie.core.models.FoodMeasure
import antoni.kalorie.core.utils.epochSecondsAsDouble
import antoni.kalorie.core.utils.instantFromEpochSeconds
import antoni.kalorie.macrokit.energyKJFromMacros
import java.time.Instant
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

@Serializable
data class FavouriteFoodDTO(
    val id: String,
    @SerialName("food_item_kind") val foodItemKind: FoodItemKind,
    @SerialName("cz_name") val czName: String,
    @SerialName("eng_name") val engName: String,
    val weight: Double,
    val date: Double,
    @SerialName("energy_kj") val energyKJ: Double? = null,
    @SerialName("calories_per_hundred_grams") val caloriesPerHundredGrams: Double,
    val fat: Double,
    @SerialName("fat_saturated") val fatSaturated: Double? = null,
    @SerialName("fat_unsaturated_fatty_acids") val fatUnsaturatedFattyAcids: Double,
    val carbohydrate: Double,
    @SerialName("carbohydrate_pure_sugar") val carbohydratePureSugar: Double,
    val fiber: Double? = null,
    val protein: Double,
    val salt: Double,
    @SerialName("favourited_at") val favouritedAt: Double,
    val portions: List<FoodPortionDTO>? = null,
    @SerialName("measure_unit") val measureUnit: String? = null,
) {

    // MARK: - Init

    constructor(item: FoodItemDomain, favouritedAt: Instant) : this(
        id = item.id,
        foodItemKind = item.kind,
        czName = item.czName,
        engName = item.engName,
        weight = item.weight,
        date = item.date.epochSecondsAsDouble(),
        energyKJ = item.energyKJ,
        caloriesPerHundredGrams = item.caloriesPerHundredGrams,
        fat = item.fat,
        fatSaturated = item.fatSaturated,
        fatUnsaturatedFattyAcids = item.fatUnsaturatedFattyAcids,
        carbohydrate = item.carbohydrate,
        carbohydratePureSugar = item.carbohydratePureSugar,
        fiber = item.fiber,
        protein = item.protein,
        salt = item.salt,
        favouritedAt = favouritedAt.epochSecondsAsDouble(),
        portions = item.portions.map(::FoodPortionDTO),
        measureUnit = item.measure.rawValue,
    )

    // MARK: - Functions

    fun asDomain(): FoodItemDomain = FoodItemDomain(
        id = id,
        kind = foodItemKind,
        czName = czName,
        engName = engName,
        weight = weight,
        date = instantFromEpochSeconds(date),
        energyKJ = energyKJ ?: energyKJFromMacros(fat = fat, carbohydrate = carbohydrate, protein = protein),
        caloriesPerHundredGrams = caloriesPerHundredGrams,
        fat = fat,
        fatSaturated = fatSaturated,
        fatUnsaturatedFattyAcids = fatUnsaturatedFattyAcids,
        carbohydrate = carbohydrate,
        carbohydratePureSugar = carbohydratePureSugar,
        fiber = fiber,
        protein = protein,
        salt = salt,
        portions = portions?.map(FoodPortionDTO::asDomain) ?: emptyList(),
        measure = measureUnit?.let(FoodMeasure::fromRawValue) ?: FoodMeasure.GRAMS,
    )
}
